File: tldrpp/tui.py
"""Terminal user interface for tldr++."""

from typing import List, Optional

from textual.app import App as TextualApp, ComposeResult
from textual.containers import Container, Horizontal, Vertical
from textual.reactive import reactive
from textual.widgets import (
    Button,
    DataTable,
    Footer,
    Header,
    Input,
    Label,
    ListView,
    ListItem,
    Static,
    TabbedContent,
    Tab,
    TextArea,
)

from tldrpp.cache import CacheManager, Page
from tldrpp.config import Config


class SearchWidget(Static):
    """Search input widget."""
    
    def __init__(self, placeholder: str = "Search commands...") -> None:
        """Initialize search widget."""
        super().__init__()
        self.placeholder = placeholder
    
    def compose(self) -> ComposeResult:
        """Compose the widget."""
        yield Input(placeholder=self.placeholder, id="search_input")


class PagesWidget(Static):
    """Pages list widget."""
    
    def __init__(self, pages: List[Page]) -> None:
        """Initialize pages widget."""
        super().__init__()
        self.pages = pages
    
    def compose(self) -> ComposeResult:
        """Compose the widget."""
        with ListView(id="pages_list"):
            for page in self.pages:
                yield ListItem(
                    Label(f"{page.name} - {page.description} ({page.platform})"),
                    id=f"page_{page.name}"
                )


class ExamplesWidget(Static):
    """Examples widget."""
    
    def __init__(self, page: Page) -> None:
        """Initialize examples widget."""
        super().__init__()
        self.page = page
    
    def compose(self) -> ComposeResult:
        """Compose the widget."""
        with Vertical():
            yield Label(f"Examples for {self.page.name}", id="examples_title")
            
            with TabbedContent():
                for i, example in enumerate(self.page.examples):
                    with Tab(f"Example {i+1}", id=f"example_{i}"):
                        yield Label(example.description, id=f"desc_{i}")
                        yield TextArea(
                            example.command,
                            read_only=True,
                            id=f"command_{i}"
                        )


class TUIApp(TextualApp):
    """Main TUI application."""
    
    CSS = """
    Screen {
        layout: vertical;
    }
    
    Header {
        dock: top;
    }
    
    Footer {
        dock: bottom;
    }
    
    #main_container {
        layout: horizontal;
        height: 1fr;
    }
    
    #left_panel {
        width: 30%;
        border: solid $primary;
        margin: 1;
    }
    
    #right_panel {
        width: 70%;
        border: solid $primary;
        margin: 1;
    }
    
    #search_input {
        margin: 1;
    }
    
    #pages_list {
        height: 1fr;
    }
    
    #examples_title {
        text-style: bold;
        margin: 1;
    }
    
    TabbedContent {
        height: 1fr;
    }
    
    Tab {
        padding: 1;
    }
    
    TextArea {
        height: 1fr;
        margin: 1;
    }
    """
    
    def __init__(self, config: Config, cache: CacheManager) -> None:
        """Initialize TUI app."""
        super().__init__()
        self.config = config
        self.cache = cache
        self.pages: List[Page] = []
        self.selected_page: Optional[Page] = None
    
    def compose(self) -> ComposeResult:
        """Compose the app."""
        yield Header()
        
        with Container(id="main_container"):
            with Vertical(id="left_panel"):
                yield SearchWidget()
                yield PagesWidget(self.pages)
            
            with Vertical(id="right_panel"):
                yield Static("Select a page to view examples", id="examples_placeholder")
        
        yield Footer()
    
    def on_mount(self) -> None:
        """Handle app mount."""
        self.title = "tldr++ - Interactive Cheat-Sheets"
        self.sub_title = "Fuzzy search and inline editing for tldr pages"
        
        # Load initial pages
        self.load_pages("")
    
    def on_input_changed(self, event: Input.Changed) -> None:
        """Handle input changes."""
        if event.input.id == "search_input":
            self.load_pages(event.value)
    
    def on_list_view_selected(self, event: ListView.Selected) -> None:
        """Handle list selection."""
        if event.list_view.id == "pages_list":
            selected_item = event.item
            if selected_item:
                page_name = selected_item.id.replace("page_", "")
                self.selected_page = next(
                    (p for p in self.pages if p.name == page_name), None
                )
                if self.selected_page:
                    self.show_examples()
    
    def load_pages(self, query: str) -> None:
        """Load pages based on search query."""
        try:
            self.pages = self.cache.search_pages(query, self.config.platforms)
            self.refresh_pages_list()
        except Exception as e:
            self.notify(f"Error loading pages: {e}", severity="error")
    
    def refresh_pages_list(self) -> None:
        """Refresh the pages list widget."""
        pages_list = self.query_one("#pages_list", ListView)
        pages_list.clear()
        
        for page in self.pages:
            item = ListItem(
                Label(f"{page.name} - {page.description} ({page.platform})"),
                id=f"page_{page.name}"
            )
            pages_list.append(item)
    
    def show_examples(self) -> None:
        """Show examples for the selected page."""
        if not self.selected_page:
            return
        
        # Replace the right panel content
        right_panel = self.query_one("#right_panel", Vertical)
        right_panel.remove_children()
        
        examples_widget = ExamplesWidget(self.selected_page)
        right_panel.mount(examples_widget)
    
    def run(self, search_query: str = "") -> None:
        """Run the TUI."""
        self.run()

File: internal/tui/tui.go
package tui

import (
	"fmt"
	"strings"

	"github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/makalin/tldrpp/internal/cache"
	"github.com/makalin/tldrpp/internal/config"
	"github.com/makalin/tldrpp/internal/types"
)

// App represents the main TUI application
type App struct {
	config      *config.Config
	cache       *cache.Manager
	state       AppState
	searchQuery string
	pages       []*types.Page
	selectedIdx int
	platforms   []string
	theme        Theme
}

// AppState represents the current state of the application
type AppState int

const (
	StateSearch AppState = iota
	StatePages
	StateExamples
	StateEdit
	StateHelp
)

// Theme represents the UI theme
type Theme struct {
	Background   lipgloss.Color
	Foreground   lipgloss.Color
	Accent       lipgloss.Color
	Success      lipgloss.Color
	Warning      lipgloss.Color
	Error        lipgloss.Color
	Border       lipgloss.Color
	Highlight    lipgloss.Color
}

// New creates a new TUI application
func New(cfg *config.Config, cacheManager *cache.Manager) *App {
	app := &App{
		config:    cfg,
		cache:     cacheManager,
		state:     StateSearch,
		platforms: cfg.Platforms,
		theme:     getTheme(cfg.Theme),
	}
	
	return app
}

// Run starts the TUI application
func (a *App) Run(searchQuery string) error {
	a.searchQuery = searchQuery
	
	// Load initial pages
	if err := a.loadPages(); err != nil {
		return fmt.Errorf("failed to load pages: %w", err)
	}

	// Create and run the bubbletea program
	p := bubbletea.NewProgram(a, bubbletea.WithAltScreen())
	_, err := p.Run()
	return err
}

// Init initializes the bubbletea model
func (a *App) Init() bubbletea.Cmd {
	return nil
}

// Update handles bubbletea updates
func (a *App) Update(msg bubbletea.Msg) (bubbletea.Model, bubbletea.Cmd) {
	switch msg := msg.(type) {
	case bubbletea.KeyMsg:
		return a.handleKeyPress(msg)
	case bubbletea.WindowSizeMsg:
		return a.handleResize(msg)
	}
	return a, nil
}

// View renders the TUI
func (a *App) View() string {
	switch a.state {
	case StateSearch:
		return a.renderSearch()
	case StatePages:
		return a.renderPages()
	case StateExamples:
		return a.renderExamples()
	case StateEdit:
		return a.renderEdit()
	case StateHelp:
		return a.renderHelp()
	default:
		return a.renderSearch()
	}
}

// handleKeyPress handles keyboard input
func (a *App) handleKeyPress(msg bubbletea.KeyMsg) (bubbletea.Model, bubbletea.Cmd) {
	switch msg.String() {
	case "ctrl+c", "q":
		return a, bubbletea.Quit
	case "?":
		if a.state == StateHelp {
			a.state = StateSearch
		} else {
			a.state = StateHelp
		}
	case "enter":
		if a.state == StateSearch {
			a.state = StatePages
		} else if a.state == StatePages {
			a.state = StateExamples
		}
	case "esc":
		switch a.state {
		case StatePages:
			a.state = StateSearch
		case StateExamples:
			a.state = StatePages
		case StateEdit:
			a.state = StateExamples
		case StateHelp:
			a.state = StateSearch
		}
	case "tab":
		if a.state == StateExamples {
			a.state = StateEdit
		}
	case "ctrl+enter":
		if a.state == StateExamples || a.state == StateEdit {
			return a.executeCommand()
		}
	case "y":
		if a.state == StateExamples || a.state == StateEdit {
			return a.copyCommand()
		}
	case "p":
		if a.state == StateExamples || a.state == StateEdit {
			return a.pasteCommand()
		}
	case "r":
		if a.state == StateSearch {
			return a.refreshCache()
		}
	case "o":
		if a.state == StateExamples {
			return a.openInPager()
		}
	case "a":
		if a.state == StatePages {
			a.toggleAllPlatforms()
		}
	case "1", "2", "3", "4", "5", "6":
		if a.state == StatePages {
			a.togglePlatform(msg.String())
		}
	case "up", "k":
		if a.selectedIdx > 0 {
			a.selectedIdx--
		}
	case "down", "j":
		if a.selectedIdx < len(a.pages)-1 {
			a.selectedIdx++
		}
	}

	return a, nil
}

// handleResize handles window resize events
func (a *App) handleResize(msg bubbletea.WindowSizeMsg) (bubbletea.Model, bubbletea.Cmd) {
	return a, nil
}

// loadPages loads pages based on current search query and platforms
func (a *App) loadPages() error {
	pages, err := a.cache.SearchPages(a.searchQuery, a.platforms)
	if err != nil {
		return err
	}
	a.pages = pages
	a.selectedIdx = 0
	return nil
}

// renderSearch renders the search interface
func (a *App) renderSearch() string {
	var content strings.Builder
	
	// Title
	title := lipgloss.NewStyle().
		Foreground(a.theme.Accent).
		Bold(true).
		Render("tldr++ - Interactive Cheat-Sheets")
	
	content.WriteString(title + "\n\n")
	
	// Search box
	searchBox := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(a.theme.Border).
		Padding(1, 2).
		Render(fmt.Sprintf("Search: %s", a.searchQuery))
	
	content.WriteString(searchBox + "\n\n")
	
	// Instructions
	instructions := lipgloss.NewStyle().
		Foreground(a.theme.Foreground).
		Render("Press Enter to search, ? for help, q to quit")
	
	content.WriteString(instructions)
	
	return content.String()
}

// renderPages renders the pages list
func (a *App) renderPages() string {
	var content strings.Builder
	
	// Header
	header := lipgloss.NewStyle().
		Foreground(a.theme.Accent).
		Bold(true).
		Render(fmt.Sprintf("Pages (%d found)", len(a.pages)))
	
	content.WriteString(header + "\n\n")
	
	// Platform filters
	platforms := lipgloss.NewStyle().
		Foreground(a.theme.Foreground).
		Render(fmt.Sprintf("Platforms: %s", strings.Join(a.platforms, ", ")))
	
	content.WriteString(platforms + "\n\n")
	
	// Pages list
	for i, page := range a.pages {
		style := lipgloss.NewStyle().Foreground(a.theme.Foreground)
		if i == a.selectedIdx {
			style = style.Background(a.theme.Highlight).Foreground(a.theme.Background)
		}
		
		pageText := fmt.Sprintf("%s - %s (%s)", page.Name, page.Description, page.Platform)
		content.WriteString(style.Render(pageText) + "\n")
	}
	
	// Footer
	footer := lipgloss.NewStyle().
		Foreground(a.theme.Foreground).
		Render("↑↓ Navigate, Enter Select, Esc Back, ? Help")
	
	content.WriteString("\n" + footer)
	
	return content.String()
}

// renderExamples renders the examples for the selected page
func (a *App) renderExamples() string {
	if len(a.pages) == 0 || a.selectedIdx >= len(a.pages) {
		return "No pages available"
	}
	
	page := a.pages[a.selectedIdx]
	var content strings.Builder
	
	// Header
	header := lipgloss.NewStyle().
		Foreground(a.theme.Accent).
		Bold(true).
		Render(fmt.Sprintf("%s - %s", page.Name, page.Description))
	
	content.WriteString(header + "\n\n")
	
	// Examples
	for i, example := range page.Examples {
		style := lipgloss.NewStyle().Foreground(a.theme.Foreground)
		if i == 0 { // Highlight first example
			style = style.Background(a.theme.Highlight).Foreground(a.theme.Background)
		}
		
		exampleText := fmt.Sprintf("%s\n  %s", example.Description, example.Command)
		content.WriteString(style.Render(exampleText) + "\n\n")
	}
	
	// Footer
	footer := lipgloss.NewStyle().
		Foreground(a.theme.Foreground).
		Render("Tab Edit, Ctrl+Enter Run, y Copy, p Paste, Esc Back")
	
	content.WriteString(footer)
	
	return content.String()
}

// renderEdit renders the placeholder editing interface
func (a *App) renderEdit() string {
	if len(a.pages) == 0 || a.selectedIdx >= len(a.pages) {
		return "No pages available"
	}
	
	page := a.pages[a.selectedIdx]
	if len(page.Examples) == 0 {
		return "No examples available"
	}
	
	example := page.Examples[0] // Use first example for now
	var content strings.Builder
	
	// Header
	header := lipgloss.NewStyle().
		Foreground(a.theme.Accent).
		Bold(true).
		Render(fmt.Sprintf("Edit: %s", example.Description))
	
	content.WriteString(header + "\n\n")
	
	// Command with placeholders
	command := example.Command
	for _, placeholder := range example.Placeholders {
		placeholderText := fmt.Sprintf("{{%s}}", placeholder.Name)
		highlighted := lipgloss.NewStyle().
			Background(a.theme.Warning).
			Foreground(a.theme.Background).
			Render(placeholderText)
		command = strings.Replace(command, placeholderText, highlighted, 1)
	}
	
	commandBox := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(a.theme.Border).
		Padding(1, 2).
		Render(command)
	
	content.WriteString(commandBox + "\n\n")
	
	// Placeholders
	if len(example.Placeholders) > 0 {
		placeholders := lipgloss.NewStyle().
			Foreground(a.theme.Foreground).
			Render("Placeholders:")
		content.WriteString(placeholders + "\n")
		
		for _, placeholder := range example.Placeholders {
			placeholderText := fmt.Sprintf("  %s (%s): %s", 
				placeholder.Name, placeholder.Type, placeholder.Default)
			content.WriteString(placeholderText + "\n")
		}
	}
	
	// Footer
	footer := lipgloss.NewStyle().
		Foreground(a.theme.Foreground).
		Render("Ctrl+Enter Run, y Copy, p Paste, Esc Back")
	
	content.WriteString("\n" + footer)
	
	return content.String()
}

// renderHelp renders the help screen
func (a *App) renderHelp() string {
	var content strings.Builder
	
	// Title
	title := lipgloss.NewStyle().
		Foreground(a.theme.Accent).
		Bold(true).
		Render("tldr++ Help")
	
	content.WriteString(title + "\n\n")
	
	// Keybindings
	keybindings := []struct {
		key, description string
	}{
		{"Enter", "Accept example / Select page"},
		{"Tab", "Edit placeholders"},
		{"Ctrl+Enter", "Run command (safe)"},
		{"y", "Copy to clipboard"},
		{"p", "Paste to terminal"},
		{"1-6", "Toggle platform filters"},
		{"a", "Toggle all platforms"},
		{"r", "Refresh cache"},
		{"o", "Open in pager"},
		{"?", "Show/hide help"},
		{"Esc", "Go back"},
		{"q", "Quit"},
	}
	
	for _, kb := range keybindings {
		key := lipgloss.NewStyle().
			Foreground(a.theme.Accent).
			Bold(true).
			Render(kb.key)
		desc := lipgloss.NewStyle().
			Foreground(a.theme.Foreground).
			Render(kb.description)
		content.WriteString(fmt.Sprintf("%-15s %s\n", key, desc))
	}
	
	// Footer
	footer := lipgloss.NewStyle().
		Foreground(a.theme.Foreground).
		Render("Press ? to close help")
	
	content.WriteString("\n" + footer)
	
	return content.String()
}

// executeCommand executes the current command
func (a *App) executeCommand() (bubbletea.Model, bubbletea.Cmd) {
	// This would execute the command
	// For now, just show a message
	return a, bubbletea.Quit
}

// copyCommand copies the current command to clipboard
func (a *App) copyCommand() (bubbletea.Model, bubbletea.Cmd) {
	// This would copy to clipboard
	// For now, just show a message
	return a, bubbletea.Quit
}

// pasteCommand pastes the current command to terminal
func (a *App) pasteCommand() (bubbletea.Model, bubbletea.Cmd) {
	// This would paste to terminal
	// For now, just show a message
	return a, bubbletea.Quit
}

// refreshCache refreshes the pages cache
func (a *App) refreshCache() (bubbletea.Model, bubbletea.Cmd) {
	// This would refresh the cache
	// For now, just reload pages
	a.loadPages()
	return a, nil
}

// openInPager opens the current page in a pager
func (a *App) openInPager() (bubbletea.Model, bubbletea.Cmd) {
	// This would open in pager
	// For now, just show a message
	return a, bubbletea.Quit
}

// toggleAllPlatforms toggles all platform filters
func (a *App) toggleAllPlatforms() {
	allPlatforms := []string{"common", "linux", "osx", "sunos", "windows", "android"}
	if len(a.platforms) == len(allPlatforms) {
		a.platforms = []string{"common"}
	} else {
		a.platforms = allPlatforms
	}
	a.loadPages()
}

// togglePlatform toggles a specific platform filter
func (a *App) togglePlatform(platformNum string) {
	platformMap := map[string]string{
		"1": "common",
		"2": "linux", 
		"3": "osx",
		"4": "sunos",
		"5": "windows",
		"6": "android",
	}
	
	platform := platformMap[platformNum]
	if platform == "" {
		return
	}
	
	// Toggle platform
	var newPlatforms []string
	found := false
	for _, p := range a.platforms {
		if p == platform {
			found = true
		} else {
			newPlatforms = append(newPlatforms, p)
		}
	}
	
	if !found {
		newPlatforms = append(newPlatforms, platform)
	}
	
	a.platforms = newPlatforms
	a.loadPages()
}

// getTheme returns the theme configuration
func getTheme(themeName string) Theme {
	switch themeName {
	case "light":
		return Theme{
			Background: lipgloss.Color("#ffffff"),
			Foreground: lipgloss.Color("#000000"),
			Accent:     lipgloss.Color("#0066cc"),
			Success:    lipgloss.Color("#00aa00"),
			Warning:    lipgloss.Color("#ffaa00"),
			Error:      lipgloss.Color("#cc0000"),
			Border:     lipgloss.Color("#cccccc"),
			Highlight:  lipgloss.Color("#e6f3ff"),
		}
	case "solarized":
		return Theme{
			Background: lipgloss.Color("#002b36"),
			Foreground: lipgloss.Color("#839496"),
			Accent:     lipgloss.Color("#268bd2"),
			Success:    lipgloss.Color("#859900"),
			Warning:    lipgloss.Color("#b58900"),
			Error:      lipgloss.Color("#dc322f"),
			Border:     lipgloss.Color("#586e75"),
			Highlight:  lipgloss.Color("#073642"),
		}
	default: // dark
		return Theme{
			Background: lipgloss.Color("#1e1e1e"),
			Foreground: lipgloss.Color("#ffffff"),
			Accent:     lipgloss.Color("#007acc"),
			Success:    lipgloss.Color("#00aa00"),
			Warning:    lipgloss.Color("#ffaa00"),
			Error:      lipgloss.Color("#cc0000"),
			Border:     lipgloss.Color("#333333"),
			Highlight:  lipgloss.Color("#2d2d30"),
		}
	}
}